io/io/src/TBufferJSON.cxx

```diff
@@ -428,23 +422,85 @@ TString TBufferJSON::ConvertToJSON(const TObject *obj, Int_t compact, const char
-//  - 2 - remove newlines
-//  - 3 - exclude spaces as much as possible
-//
-// Second digit of compact parameter defines algorithm for arrays compression
-//  - 0 - no compression, standard JSON array
-//  - 1 - exclude leading, trailing zeros, required JSROOT v5
-//  - 2 - check values repetition and empty gaps, required JSROOT v5
+/// Set level of space/newline/array compression
+/// Lower digit of compact parameter define formatting rules
+///  - 0 - no any compression, human-readable form
+///  - 1 - exclude spaces in the begin
+///  - 2 - remove newlines
+///  - 3 - exclude spaces as much as possible
+///
+/// Second digit of compact parameter defines algorithm for arrays compression
+///  - 0 - no compression, standard JSON array
+///  - 1 - exclude leading and trailing zeros
+///  - 2 - check values repetition and empty gaps
+///
+/// If third digit of compact parameter is 1, "_typename" will be skipped
 
 void TBufferJSON::SetCompact(int level)
 {
-   fCompact = level;
-   fSemicolon = (fCompact % 10 > 2) ? ":" : " : ";
-   fArraySepar = (fCompact % 10 > 2) ? "," : ", ";
+   if (level < 0)
+      level = 0;
+   fCompact = level % 10;
+   fSemicolon = (fCompact > 2) ? ":" : " : ";
+   fArraySepar = (fCompact > 2) ? "," : ", ";
+   fArrayCompact = (level / 10) % 10;
+   if (((level / 100) % 10) == 1)
+      fTypeNameTag.Clear();
+   else if (fTypeNameTag.Length() == 0)
+      fTypeNameTag = "_typename";
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Configures _typename tag in JSON structures
+/// By default "_typename" field in JSON structures used to store class information
+/// One can specify alternative tag like "$typename" or "xy", but such JSON can not be correctly used in JSROOT
+/// If empty string is provided, class information will not be stored
+
+void TBufferJSON::SetTypenameTag(const char *tag)
+{
+   if (!tag)
+      fTypeNameTag.Clear();
+   else
+      fTypeNameTag = tag;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Configures _typeversion tag in JSON
+/// One can specify name of the JSON tag like "_typeversion" ot "$tv" which will be used to store class version
+/// Such tag can be used to correctly recover objects from JSON
+/// If empty string is provided (default), class version will not be stored
+
+void TBufferJSON::SetTypeversionTag(const char *tag)
+{
+   if (!tag)
+      fTypeVersionTag.Clear();
+   else
+      fTypeVersionTag = tag;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Specify class which type information (name and version) will not be stored in JSON
+/// One can configure several such classes
+/// Allows to excluded type information for only these classes
+
+void TBufferJSON::SetSkipClassInfo(const TClass *cl)
+{
+   if (!cl)
+      return;
+   if (!fSkipClasses)
+      fSkipClasses = new TExMap();
+
+   fSkipClasses->Add(Void_Hash(cl), 777);
```

Comment:
why 777?  why pass Void_Hash(cl) instead of cl?

Reply:
> why pass Void_Hash(cl) instead of cl?

I just take same logic as in TBuffer classes to store object references

> why 777?

I need to specify any non-zero value.

But in principle - you are right. I can use plain TObjArray. 
Performance here is not critical - normally one want exclude very few classes.
To exclude all of them - one should call TBufferJSON::SetTypenameTag("")

Reply:
right I forgot TExMap is a low level interface/container.